File: repos/Djinn-Inc__djinn/sdk/src/crypto.ts
/**
 * Cryptographic primitives for the Djinn Protocol.
 *
 * - AES-256-GCM encryption via Web Crypto API (zero external dependencies)
 * - Shamir Secret Sharing over the BN254 scalar field
 *
 * The Shamir implementation matches the validator's Python crypto.py exactly.
 */

export const BN254_PRIME =
  21888242871839275222246405745257275088548364400416034343698204186575808495617n;

export interface ShamirShare {
  x: number;
  y: bigint;
}

function mod(a: bigint, p: bigint): bigint {
  const r = a % p;
  return r < 0n ? r + p : r;
}

function extendedGcd(a: bigint, b: bigint): [bigint, bigint, bigint] {
  if (a === 0n) return [b, 0n, 1n];
  const [g, x, y] = extendedGcd(mod(b, a), a);
  return [g, y - (b / a) * x, x];
}

function modInverse(a: bigint, p: bigint): bigint {
  const a2 = mod(a, p);
  const [g, x] = extendedGcd(a2, p);
  if (g !== 1n) throw new Error("Modular inverse does not exist");
  return mod(x, p);
}

function modPow(base: bigint, exp: bigint, p: bigint): bigint {
  let result = 1n;
  let b = mod(base, p);
  let e = exp;
  while (e > 0n) {
    if (e & 1n) result = mod(result * b, p);
    b = mod(b * b, p);
    e >>= 1n;
  }
  return result;
}

function getRandomFieldElement(prime: bigint): bigint {
  const bytes = new Uint8Array(32);
  for (let attempt = 0; attempt < 256; attempt++) {
    crypto.getRandomValues(bytes);
    let val = 0n;
    for (const b of bytes) {
      val = (val << 8n) | BigInt(b);
    }
    if (val < prime) return val;
  }
  throw new Error("Failed to generate unbiased random field element");
}

export function splitSecret(
  secret: bigint,
  n: number = 10,
  k: number = 7,
  prime: bigint = BN254_PRIME,
): ShamirShare[] {
  // Strict parameter validation. Without these guards:
  //   k < 2 produces a degree-0 polynomial whose shares ALL equal the
  //     secret — any single share trivially reconstructs it.
  //   k > n leaves shares that cannot be reconstructed (the threshold
  //     exceeds the share count), permanently losing the signal.
  //   non-integer / negative / NaN values silently produce malformed
  //     shares.
  // Codex audit 2026-04-15 flagged these as Critical (sdk/src/crypto.ts).
  if (!Number.isInteger(n) || n < 1) {
    throw new Error(`Shamir n must be a positive integer (got ${n})`);
  }
  if (!Number.isInteger(k) || k < 2) {
    throw new Error(
      `Shamir threshold k must be an integer >= 2 (got ${k}). ` +
        `k < 2 produces a degree-0 polynomial; one share leaks the secret.`,
    );
  }
  if (k > n) {
    throw new Error(`Shamir threshold k (${k}) cannot exceed share count n (${n})`);
  }
  if (secret >= prime) throw new Error("Secret must be < prime");
  if (secret < 0n) throw new Error("Secret must be non-negative");

  const coeffs: bigint[] = [secret];
  for (let i = 1; i < k; i++) {
    coeffs.push(getRandomFieldElement(prime));
  }
  const shares: ShamirShare[] = [];
  for (let i = 1; i <= n; i++) {
    let y = 0n;
    const x = BigInt(i);
    for (let j = 0; j < coeffs.length; j++) {
      y = mod(y + coeffs[j] * modPow(x, BigInt(j), prime), prime);
    }
    shares.push({ x: i, y });
  }
  return shares;
}

export function reconstructSecret(
  shares: ShamirShare[],
  prime: bigint = BN254_PRIME,
): bigint {
  const k = shares.length;
  let secret = 0n;
  for (let i = 0; i < k; i++) {
    const xi = BigInt(shares[i].x);
    const yi = shares[i].y;
    let numerator = 1n;
    let denominator = 1n;
    for (let j = 0; j < k; j++) {
      if (i === j) continue;
      const xj = BigInt(shares[j].x);
      numerator = mod(numerator * (0n - xj), prime);
      denominator = mod(denominator * (xi - xj), prime);
    }
    const lagrangeCoeff = mod(numerator * modInverse(denominator, prime), prime);
    secret = mod(secret + yi * lagrangeCoeff, prime);
  }
  return secret;
}

function toArrayBuffer(arr: Uint8Array): ArrayBuffer {
  return arr.buffer instanceof ArrayBuffer
    ? arr.buffer.slice(arr.byteOffset, arr.byteOffset + arr.byteLength)
    : new Uint8Array(arr).buffer as ArrayBuffer;
}

export function generateAesKey(): Uint8Array {
  const val = getRandomFieldElement(BN254_PRIME);
  return bigIntToKey(val);
}

export function toHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

const MAX_HEX_LENGTH = 131_072;

export function fromHex(hex: string): Uint8Array {
  if (hex.length > MAX_HEX_LENGTH) {
    throw new Error(`Hex string too large: ${hex.length} chars (max ${MAX_HEX_LENGTH})`);
  }
  if (hex.length % 2 !== 0) throw new Error("Hex string must have even length");
  if (hex.length > 0 && !/^[0-9a-fA-F]+$/.test(hex)) {
    throw new Error("Invalid hex characters");
  }
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) {
    bytes[i / 2] = parseInt(hex.substring(i, i + 2), 16);
  }
  return bytes;
}

export async function encrypt(
  plaintext: string,
  key: Uint8Array,
): Promise<{ ciphertext: string; iv: string }> {
  const iv = crypto.getRandomValues(new Uint8Array(12));
  const cryptoKey = await crypto.subtle.importKey(
    "raw",
    toArrayBuffer(key),
    { name: "AES-GCM" },
    false,
    ["encrypt"],
  );
  const encoded = new TextEncoder().encode(plaintext);
  const encrypted = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv: toArrayBuffer(iv) },
    cryptoKey,
    encoded,
  );
  return { ciphertext: toHex(new Uint8Array(encrypted)), iv: toHex(iv) };
}

export async function decrypt(
  ciphertext: string,
  iv: string,
  key: Uint8Array,
): Promise<string> {
  const cryptoKey = await crypto.subtle.importKey(
    "raw",
    toArrayBuffer(key),
    { name: "AES-GCM" },
    false,
    ["decrypt"],
  );
  const ivBytes = fromHex(iv);
  const ctBytes = fromHex(ciphertext);
  const decrypted = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv: toArrayBuffer(ivBytes) },
    cryptoKey,
    toArrayBuffer(ctBytes),
  );
  return new TextDecoder().decode(decrypted);
}

export function keyToBigInt(key: Uint8Array): bigint {
  let val = 0n;
  for (const b of key) val = (val << 8n) | BigInt(b);
  if (val >= BN254_PRIME) {
    throw new Error("Key value exceeds BN254 field");
  }
  return val;
}

export function bigIntToKey(val: bigint): Uint8Array {
  const key = new Uint8Array(32);
  let v = val;
  for (let i = 31; i >= 0; i--) {
    key[i] = Number(v & 0xffn);
    v >>= 8n;
  }
  return key;
}

File: repos/Djinn-Inc__djinn/validator/djinn_validator/utils/crypto.py
"""Shamir Secret Sharing and related cryptographic primitives."""

from __future__ import annotations

import secrets
from dataclasses import dataclass

# BN254 scalar field prime (same field as our ZK circuits)
BN254_PRIME = 21888242871839275222246405745257275088548364400416034343698204186575808495617


@dataclass(frozen=True)
class Share:
    """A single Shamir share: (x, y) where y = f(x) for secret polynomial f."""

    x: int
    y: int


def _mod_inv(a: int, p: int) -> int:
    """Modular multiplicative inverse using extended Euclidean algorithm."""
    if a < 0:
        a = a % p
    g, x, _ = _extended_gcd(a, p)
    if g != 1:
        raise ValueError("Modular inverse does not exist")
    return x % p


def _extended_gcd(a: int, b: int) -> tuple[int, int, int]:
    if a == 0:
        return b, 0, 1
    g, x, y = _extended_gcd(b % a, a)
    return g, y - (b // a) * x, x


def split_secret(
    secret: int,
    n: int = 10,
    k: int = 7,
    prime: int = BN254_PRIME,
) -> list[Share]:
    """Split a secret integer into n Shamir shares with threshold k.

    Args:
        secret: The secret value to split (must be < prime).
        n: Total number of shares to generate.
        k: Minimum shares needed for reconstruction.
        prime: The prime field modulus.

    Returns:
        List of n Share objects.
    """
    if secret >= prime:
        raise ValueError(f"Secret must be < {prime}")

    # Random polynomial coefficients: a_0 = secret, a_1..a_{k-1} random
    coeffs = [secret] + [secrets.randbelow(prime) for _ in range(k - 1)]

    shares = []
    for i in range(1, n + 1):
        y = 0
        for j, c in enumerate(coeffs):
            y = (y + c * pow(i, j, prime)) % prime
        shares.append(Share(x=i, y=y))

    return shares


def reconstruct_secret(
    shares: list[Share],
    prime: int = BN254_PRIME,
) -> int:
    """Reconstruct the secret from k or more Shamir shares using Lagrange interpolation.

    Args:
        shares: At least k shares from the same polynomial.
        prime: The prime field modulus.

    Returns:
        The original secret value.
    """
    k = len(shares)
    secret = 0

    for i in range(k):
        xi, yi = shares[i].x, shares[i].y
        numerator = 1
        denominator = 1

        for j in range(k):
            if i == j:
                continue
            xj = shares[j].x
            numerator = (numerator * (0 - xj)) % prime
            denominator = (denominator * (xi - xj)) % prime

        lagrange_coeff = (numerator * _mod_inv(denominator, prime)) % prime
        secret = (secret + yi * lagrange_coeff) % prime

    return secret


def generate_signal_index_shares(
    real_index: int,
    n: int = 10,
    k: int = 7,
) -> list[Share]:
    """Generate Shamir shares of a signal's real index.

    Args:
        real_index: The true index (1-10) of the signal among decoys.
        n: Total number of validator shares.
        k: Threshold for reconstruction.

    Returns:
        List of n shares encoding the real index.
    """
    if not 1 <= real_index <= 10:
        raise ValueError(f"Index must be in [1, 10], got {real_index}")
    return split_secret(real_index, n, k)
